File: src/main/java/io/pageobject/generator/attribute/Attributes.java
package io.pageobject.generator.attribute;

import org.jsoup.nodes.Element;

import static com.google.common.base.Strings.isNullOrEmpty;
import static io.pageobject.generator.Expressions.allowIndexExpressionOrNoExpressions;
import static io.pageobject.generator.Expressions.hasExpression;

public class Attributes {

    public static String getNormalizedAttributeValue(Element element, String attributeKey) {
        return getNormalizedAttribute(element, attributeKey).value;
    }

    private static Attribute getNormalizedAttribute(Element element, String attributeKey) {
        Attribute attribute = new Attribute();
        String[] variants = {attributeKey, attributeKey.replaceAll("-", ":"), attributeKey.replaceAll("-", "_")};

        for (String variant : variants) {
            attribute = tryVariant(element, variant);

            if (attribute.isPresent) {
                break;
            }
        }

        return attribute;
    }

    private static Attribute tryVariant(Element element, String attributeKey) {
        Attribute attribute = new Attribute();

        String[] keys = {attributeKey,
                         "data-" + attributeKey,
                         "data_" + attributeKey,
                         "data:" + attributeKey,
                         "x-" + attributeKey,
                         "x_" + attributeKey,
                         "x:" + attributeKey};

        for (String key : keys) {
            attribute.key = key;
            attribute.value = element.attr(key);
            attribute.isPresent = element.hasAttr(key);

            if (attribute.isPresent) {
                break;
            }
        }

        return attribute;
    }

    public static boolean hasNormalizedAttribute(Element element, String attributeKey) {
        return getNormalizedAttribute(element, attributeKey).isPresent;
    }

    public static String getNormalizedAttributeKey(Element element, String attributeKey) {
        Attribute attribute = getNormalizedAttribute(element, attributeKey);
        return isNullOrEmpty(attribute.value) ? null : attribute.key;
    }

    public static boolean hasRepeaterAttribute(Element element) {
        return !isNullOrEmpty(getRepeaterAttribute(element));
    }

    public static String getRepeaterAttribute(Element element) {
        String repeatExpression = getNormalizedAttributeValue(element, HtmlAttributes.NG_REPEAT.attributeName());
        if (isNullOrEmpty(repeatExpression)) {
            repeatExpression = getNormalizedAttributeValue(element, HtmlAttributes.NG_REPEAT_START.attributeName());
        }

        return repeatExpression;
    }

    public static boolean hasAngular2EventBindingAttribute(Element element, String event) {
        return !isNullOrEmpty(getAngular2EventBindingAttribute(element, event));
    }

    public static String getAngular2EventBindingAttribute(Element element, String event) {
        String value = element.attr("(" + event + ")");
        if (isNullOrEmpty(value)) {
            value = element.attr("on-" + event);
        }

        return value;
    }

    public static String extractAttributeValue(Element element, HtmlAttributes htmlAttribute) {
        String attributeValue = getNormalizedAttributeValue(element, htmlAttribute.attributeName());
        return hasExpression(attributeValue) ? null : attributeValue;
    }

    public static String extractAttributeValueAllowingIndexExpression(Element element, HtmlAttributes htmlAttribute) {
        String attributeValue = getNormalizedAttributeValue(element, htmlAttribute.attributeName());
        return allowIndexExpressionOrNoExpressions(attributeValue);
    }

    private static class Attribute {
        String key;
        String value;
        boolean isPresent;
    }

}

File: lib/dom/attributes.js
'use strict';

const _ = require('lodash');
const hasExpression = require('../expressions').hasExpression;
const allowIndexExpressionOrNoExpressions = require('../expressions').allowIndexExpressionOrNoExpressions;
const HtmlAttribute = require('./html-attribute');

class Attributes {

  static getAttribute(element, key) {
    const attributes = element.attribs;

    if (attributes) {
      for (let attribute in attributes) {
        if (attribute.toUpperCase() === key.toUpperCase()) {
          return attributes[attribute];
        }
      }
    }

    return "";
  }

  static hasAttribute(element, key) {
    const attributes = element.attribs;

    if (attributes) {
      for (let attribute in attributes) {
        if (attribute.toUpperCase() === key.toUpperCase()) {
          return true;
        }
      }
    }

    return false;
  }

  static getNormalizedAttributeValue(element, attributeKey) {
    return getNormalizedAttribute(element, attributeKey).value;
  }

  static hasNormalizedAttribute(element, attributeKey) {
    return getNormalizedAttribute(element, attributeKey).isPresent;
  }

  static getNormalizedAttributeKey(element, attributeKey) {
    let attribute = getNormalizedAttribute(element, attributeKey);
    return _.isEmpty(attribute.value) ? null : attribute.key;
  }

  static hasRepeaterAttribute(element) {
    return !_.isEmpty(Attributes.getRepeaterAttribute(element));
  }

  static getRepeaterAttribute(element) {
    let repeatExpression = Attributes.getNormalizedAttributeValue(element, HtmlAttribute.NG_REPEAT.attributeName);
    if (_.isEmpty(repeatExpression)) {
      repeatExpression = Attributes.getNormalizedAttributeValue(element, HtmlAttribute.NG_REPEAT_START.attributeName);
    }

    return repeatExpression;
  }

  static hasAngular2EventBindingAttribute(element, event) {
    return !_.isEmpty(Attributes.getAngular2EventBindingAttribute(element, event));
  }

  static getAngular2EventBindingAttribute(element, event) {
    let value = Attributes.getAttribute(element, '(' + event + ')');
    if (_.isEmpty(value)) {
      value = Attributes.getAttribute(element, 'on-' + event);
    }

    return value;
  }

  static extractAttributeValue(element, htmlAttribute) {
    let attributeValue = Attributes.getNormalizedAttributeValue(element, htmlAttribute.attributeName);
    return hasExpression(attributeValue) ? null : attributeValue;
  }

  static extractAttributeValueAllowingIndexExpression(element, htmlAttribute) {
    let attributeValue = Attributes.getNormalizedAttributeValue(element, htmlAttribute.attributeName);
    return allowIndexExpressionOrNoExpressions(attributeValue);
  }
}

function getNormalizedAttribute(element, attributeKey) {
  let attribute;
  let variants = [attributeKey, attributeKey.replace(/-/g, ":"), attributeKey.replace(/-/g, "_")];

  for (let variant of variants) {
    attribute = tryVariant(element, variant);

    if (attribute.isPresent) {
      break;
    }
  }

  return attribute;
}

function tryVariant(element, attributeKey) {
  let attribute = {};
  let keys = [attributeKey,
    "data-" + attributeKey,
    "data_" + attributeKey,
    "data:" + attributeKey,
    "x-" + attributeKey,
    "x_" + attributeKey,
    "x:" + attributeKey];

  for (let key of keys) {
    attribute.key = key;
    attribute.value = Attributes.getAttribute(element, key);
    attribute.isPresent = Attributes.hasAttribute(element, key);

    if (attribute.isPresent) {
      break;
    }
  }

  return attribute;
}

module.exports = Attributes;